[ROCm] Enable Triton ScaledMM fallback + kernel selection fix

shivampr opened the pull request:
<h3><strong>Purpose</strong></h3>
<p>Fixes #14397 — <code inline="">triton_scaled_mm</code> was never used on ROCm due to missing dispatch and checks.<br>
This PR:</p>
<ul>
<li>
<p>Enables <strong>Triton fallback</strong> for ROCm when AITriton is unavailable</p>
</li>
<li>
<p>Adds Triton fallback after CUTLASS on CUDA</p>
</li>
<li>
<p>Implements <code inline="">is_supported()</code> checks for kernel selection</p>
</li>
<li>
<p>Adds a lightweight integration test validating ROCm dispatch logic</p>
</li>
</ul>
<hr>
<h3><strong>Test Plan</strong></h3>
<h4><strong>1. Mocked test (no GPU)</strong></h4>
<pre><code class="language-bash">python3 mini_tests/select_triton_rocm.py
</code></pre>
<p><strong>Result</strong></p>
<pre><code>Selected kernel: TritonScaledMMLinearKernel
OK: TritonScaledMMLinearKernel chosen on ROCm fallback.
</code></pre>
<h4><strong>2. MI300X (ROCm 7.0, vLLM built from this PR)</strong></h4>
<p><strong>(a) Triton kernel functional test</strong></p>
<pre><code>max_abs_err≈2.5e-01, max_rel_err≈3.9e-03
</code></pre>
<p><strong>(b) OpenAI-compatible API test</strong></p>
<pre><code class="language-bash">python3 -m vllm.entrypoints.openai.api_server \
  --model Qwen/Qwen2.5-0.5B-Instruct --dtype bfloat16 --host 0.0.0.0 --port 8000
</code></pre>
<p>Then:</p>
<pre><code class="language-bash">curl http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{"model":"Qwen/Qwen2.5-0.5B-Instruct","messages":[{"role":"user","content":"Say hi from MI300X."}]}'
</code></pre>
<p><strong>Response</strong></p>
<pre><code>"Hello! How can I assist you today?"
</code></pre>
<p>Confirms successful end-to-end inference on ROCm.</p>

shivampr committed:
feat(rocm): enable TritonScaledMM fallback on ROCm; add CUDA fallback entry

Signed-off-by: Shivam <shivampr.dev@gmail.com>
Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr commented on vllm/model_executor/layers/quantization/kernels/scaled_mm/triton.py:
```diff
@@ -4,39 +4,72 @@
+    def is_supported(
+        cls, compute_capability: int | None = None
+    ) -> tuple[bool, str | None]:
+        if current_platform.is_rocm() or current_platform.is_cuda():
```
Thanks, addressed this.

shivampr commented on vllm/model_executor/layers/quantization/kernels/scaled_mm/triton.py:
```diff
@@ -4,39 +4,72 @@
         return True, None
 
     def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
-        super().process_weights_after_loading(layer)
```
Triton no longer inherits from `CutlassScaledMMLinearKernel`, so I’m not calling the CUTLASS-specific `process_weights_after_loading` on the Triton path.

In this implementation I re-apply only the transforms Triton needs:

1. Transpose `w_q` to match Triton’s expected layout.
2. For static input scales, collapse to a scalar and drop input zp/azp_adj.

On CUDA, when Triton is selected (i.e., CUTLASS is unavailable), the same Triton-specific `process_weights_after_loading` runs, so behavior remains correct without CUTLASS assumptions.

ProExpertProg commented on vllm/model_executor/layers/quantization/kernels/scaled_mm/__init__.py:
```diff
@@ -68,6 +68,13 @@ def choose_scaled_mm_linear_kernel(
+            supported, reason = kernel.is_supported(compute_capability)
+            if not supported:
+                failure_reasons.append(f" {kernel.__name__}: {reason}")
+                continue
```
We should just always use `is_supported` instead of `compute_capability` like done in my draft PR #19434, could you make that change here?

ProExpertProg commented on mini_tests/select_triton_rocm.py:
```diff
@@ -0,0 +1,65 @@
+# SPDX-License-Identifier: Apache-2.0
```
Can you make this a proper pytest-bases test in `tests/...`?

ProExpertProg commented:
Is this ready for review again?

ProExpertProg commented on vllm/model_executor/layers/quantization/kernels/scaled_mm/aiter.py:
```diff
@@ -57,6 +57,14 @@ class AiterScaledMMLinearKernel(CutlassScaledMMLinearKernel):
         return 90
 
+    @classmethod
+    def is_supported(
```
Let's move the checks from `can_implement` into `is_supported` here

ProExpertProg reviewed (commented):
Just one note about `is_supported`

shivampr committed:
Refactor: Make is_supported() abstract and move platform checks

Make is_supported() abstract in base class, remove get_min_capability(),
and implement is_supported() in all kernels. Move platform checks from
can_implement() to is_supported() in AiterScaledMMLinearKernel. Add
CPU-compatible tests for kernel selection validation.

Signed-off-by: Shivam <shivamprasad91@gmail.com>

tjtanaa commented on tests/kernels/quantization/test_scaled_mm_kernel_selection.py:
```diff
@@ -0,0 +1,97 @@
+    """Test that all kernel classes implement is_supported() method."""
+    kernels = [
+        CPUScaledMMLinearKernel,
+        CutlassScaledMMLinearKernel,
```
@shivampr Just curious, why `AiterScaledMMLinearKernel` is not included in the test?

ProExpertProg commented:
@shivampr could you report lm-eval results for a model that uses int8 Triton scaled mm to check this works?

shivampr committed:
Fix CPU tests: add AiterScaledMMLinearKernel test, handle Mantis decord dependency for CI failure

Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr commented:
@ProExpertProg 

I verified the ROCm int8 Triton ScaledMM path using `lm-eval` with vLLM.
```
Environment:
- Base image: RunPod ROCm vLLM 0.9.2 / ROCm 7.0
- Backend: `VLLM_TARGET_DEVICE=rocm`
- Aiter disabled so we exercise Triton fallback:
  ```bash
  export VLLM_TARGET_DEVICE=rocm
  export VLLM_DISABLED_KERNELS="AiterScaledMMLinearKernel"
```

Command : 
```
lm_eval \
  --model vllm \
  --model_args pretrained="RedHatAI/Meta-Llama-3.1-8B-Instruct-quantized.w8a8",dtype=auto,tensor_parallel_size=1,max_model_len=4096,gpu_memory_utilization=0.9,trust_remote_code=true \
  --tasks gsm8k \
  --num_fewshot 5 \
  --limit 50 \
  --batch_size auto

```

Form logs : 

```
INFO ... Automatically detected platform rocm.
INFO ... Using TritonScaledMMLinearKernel for CompressedTensorsW8A8Int8
```
`lm-eval` results : 

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.70|±  |0.0655|
|     |       |strict-match    |     5|exact_match|↑  | 0.64|±  |0.0686|

shivampr committed:
Fix YAML: quote Mantis install command with || operator in CPU test

Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr committed:
Fix ruff : split long comment in test file

Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr committed:
Merge branch 'main' into rocm-triton-fallback
